src/Microsoft.ML.Tokenizers/Model/SentencePieceUnigramModel.cs

```diff
@@ -0,0 +1,1399 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Sentencepiece;
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.ML.Tokenizers
+{
+    internal sealed class SentencePieceUnigramModel : SentencePieceBaseModel
+    {
+        private readonly SortedDictionary<string, int> _vocab;
+        private readonly (string Piece, float Score, ModelProto.Types.SentencePiece.Types.Type Type)[] _vocabReverse;
+        private readonly DoubleArrayTrie _trie;
+        private readonly float _minScore;
+        private readonly float _maxScore;
+        private const float UnkPenalty = 10.0f;
+
+        public SentencePieceUnigramModel(ModelProto modelProto, bool addBos, bool addEos, IReadOnlyDictionary<string, int>? specialTokens = null) : base(modelProto, addBos, addEos, specialTokens)
+        {
+            _vocab = new SortedDictionary<string, int>(OrdinalUtf8StringComparer.Instance);
+            _vocabReverse = new (string Piece, float Score, ModelProto.Types.SentencePiece.Types.Type Type)[modelProto.Pieces.Count];
```

Comment:
It looks like this `ModelProto` is deserialized from a stream.  Who validates that it has legitimate data?  For instance - that the count isn't ridiculously high to cause large allocation, or that the count is sufficent to hold `BeginningOfSentenceId` and `EndOfSentenceId` below?

Reply:
We discussed this offline. In summary, we have a generated Protobuf specification that dictates how the data is deserialized. If the data does not conform to the specification, it will fail to load. Validation is handled by the Protobuf compiler during the deserialization process.